### ISSUE TITLE
Extract mask from Onnx, using cv2 dnn

### ISSUE THREAD

**Issue by davidvct** [ultralytics/ultralytics#4844]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I trained a Yolo v8 model and exported as ONNX for Instance Segmentation. I try to load it using cv2 dnn, but unsure how to load the mask from the output.

Here is a sample code I used (created by someone else):
```
import cv2
import numpy as np
from ultralytics.utils.ops import process_mask_upsample

model_path =r"./best.onnx"
image_path = r"./image01.png"

net = cv2.dnn.readNet(model_path)

# Read the input image
img = cv2.imread(image_path)
img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)

# Resize the image to the desired size (640,640) for inference
img = cv2.resize(img, (640, 640))

# Change the order of image dimensions from (640,640,3) to (3,640,640)
img = img.transpose(2, 0, 1)

# Add an extra dimension at index 0
img = img.reshape(1, 3, 640, 640)

# Scale pixel values to the range 0-1
img = img / 255.0

# Set the input for the neural network
net.setInput(img)

# Run the inference
out = net.forward()

print(type(out))
print(out.shape)

result = out[0]
result = result.transpose() 
```
print output:
```
<class 'numpy.ndarray'>
(1, 32, 160, 160)
```

I found most people use onnxruntime for this. But if I am to use cv2 dnn, how do I extract the mask as well as the confidence, class, and bbox?

**Comment by glenn-jocher** [comment]
@davidvct hello,

Thank you for using our repository and for taking the time to understand more about ONNX models and instance segmentation.

Your model's output shape suggests that it returns 32 feature maps, each of size 160x160. Typically, for instance segmentation tasks using YOLO-based architectures like YOLOv8, there are multiple outputs: a detection output and a segmentation mask output. 

In your case, the mask seems to come directly as an output, which might make the extraction a bit different than in a standard YOLO setup. In a standard YOLO setup, the output tensor would usually be in the shape [batch_size, num_anchors * (num_classes + 5 + num_masks), grid_size, grid_size]. The '+5' includes four bounding box offsets and an objectness score, while 'num_masks' corresponds to the masks for the instance segmentation.

Based on your code, it seems like you might be getting just the segmentation mask output. However, it's hard to be certain without knowing more about your specific model and how it was designed and trained.

Using cv2 dnn to get the mask could be a bit more complex than using onnxruntime as it requires manual handling. You might have to first extract the mask from the output tensor, then post-process it to recover the original size scale and semantics. The function `process_mask_upsample()` from Ultralytics utils.ops module that you referenced does the segmentation mask upsample operation, which could be directly applied to the segmentation output.

As for extracting bounding boxes, class scores, and objectness scores, these would typically come from the remaining part of the output tensor, after excluding the segmentation maps. You would usually need to apply an algorithm to decode the bounding boxes and class scores using the relevant anchors and other hyperparameters. Unfortunately, I can't provide more specifics without more information about your model and its design. 

I hope this provides some direction. If you have further questions, please don't hesitate to ask. Thanks for your interest in our work and ongoing support of the YOLO community.
